### PR TITLE
Combat tracker

### PR DESCRIPTION
**Pull request by Ampersandrew**
I have most of this written. I think it's ready for incorporation. Still needs other bits done, but wht's here is mostly working.

### CODE REVIEW

File: module/helpers/hand.mjs [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    this.override = -1;
+    this.initiative = -1;
+  }
+
+  static fromObject(proto) {
+    const isHostile =
+      typeof proto !== 'undefined' &&
+      typeof proto.isHostile !== 'undefined' &&
+      typeof proto.isHostile === 'boolean' &&
+      proto.isHostile;
```
Comment (TheSpyder) [inline module/helpers/hand.mjs]: There are probably a few areas we could make fancy improvements, but this is an obvious one 😂 
```suggestion
    const isHostile = Boolean(proto?.isHostile);
```

**Review by TheSpyder — DISMISSED** [review]
Some of it is a bit big to review, but I think it looks reasonable enough that we should run it on Matt's server and see what we think

File: module/helpers/hand.mjs (hunk 1 above)
Comment (Ampersandrew) [reply module/helpers/hand.mjs]: Yes, if you look at the commit called tidy up flag access you can see me tidy up lots of instances of that kind of thing.

**Review by TheSpyder — DISMISSED** [review]

**Comment by haxxxton** [comment]
@Ampersandrew, im logged in as the GM, and am getting errors when trying to draw cards for my actors.

```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
    at DeadlandsCombatTracker._onCombatantControl
```

Appears the `game['deadlands-classic'].socket` is undefined when trying to `drawCard` (line 236 of dlc-combat-tracker.mjs).

Am i doing something wrong?

**Comment by Ampersandrew** [comment]
I might not have committed all the files needed. Let me have a look.


-- 
  Andrew Wilson
  ***@***.***


On Sun, 8 Oct 2023, at 6:03 AM, haxxxton wrote:
> 
> 
> @Ampersandrew <https://github.com/Ampersandrew>, im logged in as the GM, and am getting errors when trying to draw cards for my actors.
> 
> `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
>     at DeadlandsCombatTracker._onCombatantControl
`

**Comment by Ampersandrew** [comment]
Ah no. There 's a module you have to install to get the sockets working. Sorry I forgot to mention it.

Install socketlib, that'll sort it out. I was going to reimplement this, but after I dissected it, it was just
going to waste time, so I decided to use the library.

-- 
  Andrew Wilson
  ***@***.***


On Sun, 8 Oct 2023, at 6:03 AM, haxxxton wrote:
> 
> 
> @Ampersandrew <https://github.com/Ampersandrew>, im logged in as the GM, and am getting errors when trying to draw cards for my actors.
> 
> `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
>     at DeadlandsCombatTracker._onCombatantControl
`

**Comment by Ampersandrew** [comment]
Background. using cards in any way affects the combat. Only GMs can modify the
combat. Players can modify their own participant in the combat (the combatant).
I wrote it and checked it out as I was going, I had most of it working using the GM.

Then I tried it logged in as a player and couldn't draw or use cards, because doing
that affected the combatant order which is a combat function.

I made every thing use a socket so that even the GMs client asks "A connected
GM" to make the change. This will still be done on their system, but will go through
a socket.

There are some changes that only a GM should legitimately make, the buttons
that invoke those functions are not shown to connected players, only to connected
GMs.

-- 
  Andrew Wilson
  ***@***.***


On Sun, 8 Oct 2023, at 3:58 PM, Andrew Wilson wrote:
> Ah no. There 's a module you have to install to get the sockets working. Sorry I forgot to mention it.
> 
> Install socketlib, that'll sort it out. I was going to reimplement this, but after I dissected it, it was just
> going to waste time, so I decided to use the library.
> 
> -- 
>   Andrew Wilson
>   ***@***.***
> 
> 
> On Sun, 8 Oct 2023, at 6:03 AM, haxxxton wrote:
>> 
>> 
>> @Ampersandrew <https://github.com/Ampersandrew>, im logged in as the GM, and am getting errors when trying to draw cards for my actors.
>> 
>> `Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'executeAsGM')
>>     at DeadlandsCombatTracker._onCombatantControl
`

**Comment by haxxxton** [comment]
@Ampersandrew is socketlib an npm package? how is that being included in the game system? (is there a build step that isnt in the package.json?), or should i be copying over the node_modules folder to the system folder in fftv?

**Comment by Ampersandrew** [comment]
> @Ampersandrew is socketlib an npm package? how is that being included in the game system? (is there a build step that isnt in the package.json?), or should i be copying over the node_modules folder to the system folder in fftv?

It's a foundry extension module, you install it with the Foundry application. When you fire Foundry up it gives you a screen with three options. Game worlds, Game systems and Add-on Modules. It probably should be in the package.json somewhere that the game system depends on the module. I'll have a look.

**Comment by Ampersandrew** [comment]
I have cloned the socket stuff into our system, and brought it up to our coding standards. This is allowed by its original MIT license. This means that we no longer need to install socketlib for our system to run.

**Review by TheSpyder — APPROVED** [review]
If there was some pre-requisite steps for installation that would've been fine, rather than vendoring the code into our project, but we can clean that up later

**Comment by Ampersandrew** [comment]
I had originally started cleaning it, I was interested to see how it works, and then decided I should just use it. That's what I originally put up.  Since I didn't know how to tell Foundry that it needs the library and it doesn't fall over in a nice way, it wasn't obvious to James why it wasn't running.

I mentioned on the #system-developer discord channel that I was planning have the system use an external module. No one there thought it was a good idea to have the system dependent on someone else's code. I asked how to tell Foundry that the system needs a module to run, they universally thought that was a bad idea.

Based on that, I cleaned it up and added it to our system.

**Comment by TheSpyder** [comment]
Ah that makes sense, thank you!

File: system.json [hunk 2]
```diff
@@ -48,10 +48,8 @@
...
   "gridDistance": 2,
   "gridUnits": "yrd",
-  "primaryTokenAttribute": "oWind",
-  "secondaryTokenAttribute": "oStrain",
-  "url": "https://github.com/Dulux-Oz/DeadlandsClassic",
-  "manifest": "https://raw.githubusercontent.com/Dulux-Oz/DeadlandsClassic/master/system.json",
-  "download": "https://github.com/Dulux-Oz/DeadlandsClassic/releases/download/0.1/deadlands_classic.zip",
+  "url": "https://github.com/Ampersandrew/deadlands-classic",
+  "manifest": "https://raw.githubusercontent.com/Ampersandrew/deadlands-classic/CombatTracker/system.json",
+  "download": "https://github.com/Ampersandrew/deadlands-classic/releases/download/0.1/deadlands_classic.zip",
```
Comment (TheSpyder) [inline system.json:53]: Now that I know what I'm looking at, these values are pointing to your github repo not this repo 🤔

Comment (Ampersandrew) [reply system.json:53]: I was trying to get the thing to install from github. That was part of me trying to get it to work. It only does that when your module is registered with Foundry. I forgot this was in here. It should be backed out.

Comment (TheSpyder) [reply system.json:53]: After I wrote that I realised I should be using a local checkout of the code, not telling it to install from the `system.json` link. So there's no rush, bigger problems to fix than this right now 😅

**Pull request merged**